cassandra-five-zero-bridge/src/main/java/org/apache/cassandra/spark/reader/SSTableReader.java

```diff
@@ -0,0 +1,898 @@
+    }
+
+    /**
+     * Merge all the partition key filters to give the token range we care about.
+     * If no partition key filters, then use the Spark worker token range.
+     *
+     * @param sparkRangeFilter    optional spark range filter
+     * @param partitionKeyFilters list of partition key filters
+     * @return the token range we care about for this Spark worker
+     */
+    public static Optional<TokenRange> extractRange(@Nullable SparkRangeFilter sparkRangeFilter,
+                                                    @NotNull List<PartitionKeyFilter> partitionKeyFilters)
+    {
+        Optional<TokenRange> partitionKeyRange = partitionKeyFilters.stream()
+                                                                    .map(PartitionKeyFilter::tokenRange)
+                                                                    .reduce(TokenRange::merge);
+        return partitionKeyRange.isPresent()
+                ? partitionKeyRange
+                : Optional.ofNullable(sparkRangeFilter != null ? sparkRangeFilter.tokenRange() : null);
+    }
+
+    /**
+     * Read Data.db offsets by binary searching Summary.db into Index.db, then reading offsets in Index.db
+     *
+     * @param indexSummary Summary.db index summary
+     * @param range        token range we care about for this Spark worker
+     */
+    private void readOffsets(IndexSummary indexSummary, TokenRange range)
+    {
+        try
+        {
+            if (indexSummary != null)
+            {
+                // BIG format
+                // If start is null we failed to find an overlapping token in the Index.db file,
+                // this is unlikely as we already pre-filter the SSTable based on the start-end token range.
+                // But in this situation we read the entire Data.db file to be safe, even if it hits performance.
+                startOffset = IndexDbUtils.findDataDbOffset(indexSummary, range, metadata.partitioner, ssTable, stats);
+                if (startOffset == null)
+                {
+                    LOGGER.error("Failed to find Data.db start offset, performance will be degraded sstable='{}'", ssTable);
+                }
+            }
+            else
+            {
+                // BTI format
+                try (InputStream primaryIndex = ssTable.openPrimaryIndexStream())
+                {
+                    File file = new File(ssTable.getDataFileName());
+                    long size = ssTable.length(FileType.PARTITIONS_INDEX);
+                    try (ReadOnlyInputStreamFileChannel fileChannel = new ReadOnlyInputStreamFileChannel((BufferingInputStream<?>) primaryIndex, size);
+                         ChannelProxy proxy = new ChannelProxy(file, fileChannel);
+                         FileHandle fileHandle = new FileHandle.Builder(file).complete(f -> proxy);
+                         PartitionIndex partitionIndex = PartitionIndex.load(fileHandle, metadata.partitioner, false);
```

Comment:
Disabling preloading makes sense to avoid downloading data over the wire. Maybe add a comment for posterity.